File: code/backend/python/app/adapters/rest/story/story_crud_admin_controller.py
"""
StoryCrudAdminController — REST controller for Step 17 admin CRUD endpoints.
Routes:
  POST   /api/admin/stories                           → create story
  PUT    /api/admin/stories/{uuidStory}                → update story
  GET    /api/admin/stories/{uuidStory}/{entityType}   → list entities
  POST   /api/admin/stories/{uuidStory}/{entityType}   → create entity
  GET    /api/admin/stories/{uuidStory}/{entityType}/{entityUuid} → get entity
  PUT    /api/admin/stories/{uuidStory}/{entityType}/{entityUuid} → update entity
  DELETE /api/admin/stories/{uuidStory}/{entityType}/{entityUuid} → delete entity
"""
from fastapi import APIRouter, Path, HTTPException, Body, Request
from typing import Dict, Any, List
from app.core.ports.story.story_crud_port import StoryCrudPort


class StoryCrudAdminController:
    def __init__(self, crud_port: StoryCrudPort):
        self.crud_port = crud_port
        self.router = APIRouter(prefix="/api/admin/stories", tags=["Admin Story CRUD"])

        # Story-level
        self.router.add_api_route("", self.create_story_route, methods=["POST"], status_code=201)
        self.router.add_api_route("/{uuidStory}", self.update_story_route, methods=["PUT"])

        # Entity CRUD
        self.router.add_api_route("/{uuidStory}/{entityType}", self.list_entities, methods=["GET"])
        self.router.add_api_route("/{uuidStory}/{entityType}", self.create_entity, methods=["POST"], status_code=201)
        self.router.add_api_route("/{uuidStory}/{entityType}/{entityUuid}", self.get_entity, methods=["GET"])
        self.router.add_api_route("/{uuidStory}/{entityType}/{entityUuid}", self.update_entity, methods=["PUT"])
        self.router.add_api_route("/{uuidStory}/{entityType}/{entityUuid}", self.delete_entity, methods=["DELETE"])

    # === Story-level ===

    async def create_story_route(self, req: Request, data: Dict[str, Any] = Body(default=None)):
        self._require_admin(req)
        if not data:
            raise HTTPException(status_code=400, detail={"error": "EMPTY_DATA", "message": "Request body required"})
        result = self.crud_port.create_story(data)
        if result is None:
            raise HTTPException(status_code=400, detail={"error": "INVALID_DATA", "message": "Could not create story"})
        return result

    async def update_story_route(self, req: Request, uuidStory: str = Path(...), data: Dict[str, Any] = Body(default=None)):
        self._require_admin(req)
        if not data:
            raise HTTPException(status_code=400, detail={"error": "EMPTY_DATA", "message": "Request body required"})
        result = self.crud_port.update_story(uuidStory, data)
        if result is None:
            raise HTTPException(status_code=404, detail={"error": "STORY_NOT_FOUND", "message": f"No story: {uuidStory}"})
        return result

    # === Entity CRUD ===

    async def list_entities(self, req: Request, uuidStory: str = Path(...), entityType: str = Path(...)):
        self._require_admin(req)
        result = self.crud_port.list_entities(uuidStory, entityType)
        if result is None:
            raise HTTPException(status_code=404, detail={"error": "STORY_NOT_FOUND", "message": f"No story: {uuidStory}"})
        return result

    async def create_entity(self, req: Request, uuidStory: str = Path(...), entityType: str = Path(...), data: Dict[str, Any] = Body(default=None)):
        self._require_admin(req)
        if not data:
            raise HTTPException(status_code=400, detail={"error": "EMPTY_DATA", "message": "Request body required"})
        result = self.crud_port.create_entity(uuidStory, entityType, data)
        if result is None:
            raise HTTPException(status_code=404, detail={"error": "STORY_NOT_FOUND", "message": f"No story: {uuidStory}"})
        return result

    async def get_entity(self, req: Request, uuidStory: str = Path(...), entityType: str = Path(...), entityUuid: str = Path(...)):
        self._require_admin(req)
        result = self.crud_port.get_entity(uuidStory, entityType, entityUuid)
        if result is None:
            raise HTTPException(status_code=404, detail={"error": "ENTITY_NOT_FOUND", "message": f"Not found: {entityUuid}"})
        return result

    async def update_entity(self, req: Request, uuidStory: str = Path(...), entityType: str = Path(...), entityUuid: str = Path(...), data: Dict[str, Any] = Body(default=None)):
        self._require_admin(req)
        if not data:
            raise HTTPException(status_code=400, detail={"error": "EMPTY_DATA", "message": "Request body required"})
        result = self.crud_port.update_entity(uuidStory, entityType, entityUuid, data)
        if result is None:
            raise HTTPException(status_code=404, detail={"error": "ENTITY_NOT_FOUND", "message": f"Not found: {entityUuid}"})
        return result

    async def delete_entity(self, req: Request, uuidStory: str = Path(...), entityType: str = Path(...), entityUuid: str = Path(...)):
        self._require_admin(req)
        deleted = self.crud_port.delete_entity(uuidStory, entityType, entityUuid)
        if not deleted:
            raise HTTPException(status_code=404, detail={"error": "ENTITY_NOT_FOUND", "message": f"Not found: {entityUuid}"})
        return {"status": "DELETED", "uuid": entityUuid, "entityType": entityType}

    def _require_admin(self, request: Request):
        role = request.state.role if hasattr(request.state, "role") else None
        if role != "ADMIN":
            raise HTTPException(status_code=403, detail={"error": "FORBIDDEN", "message": "Insufficient permissions"})

File: code/backend/php/src/Adapter/Rest/Story/StoryCrudAdminController.php
<?php

declare(strict_types=1);

namespace Games\Paths\Adapter\Rest\Story;

use Games\Paths\Core\Port\Story\StoryCrudPort;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

/**
 * StoryCrudAdminController — REST controller for Step 17 admin CRUD endpoints.
 */
class StoryCrudAdminController
{
    private StoryCrudPort $crudPort;

    public function __construct(StoryCrudPort $crudPort)
    {
        $this->crudPort = $crudPort;
    }

    // POST /api/admin/stories
    public function createStory(Request $request, Response $response, array $args): Response
    {
        if (!$this->isAdmin($request)) {
            return $this->forbidden($response);
        }
        $data = $request->getParsedBody();
        if (empty($data)) {
            return $this->jsonError($response, 400, 'EMPTY_DATA', 'Request body required');
        }
        $result = $this->crudPort->createStory($data);
        if ($result === null) {
            return $this->jsonError($response, 400, 'INVALID_DATA', 'Could not create story');
        }
        $response->getBody()->write(json_encode($result));
        return $response->withStatus(201)->withHeader('Content-Type', 'application/json');
    }

    // GET /api/admin/stories/{uuidStory}
    public function getStory(Request $request, Response $response, array $args): Response
    {
        if (!$this->isAdmin($request)) {
            return $this->forbidden($response);
        }
        $result = $this->crudPort->getStory($args['uuidStory']);
        if ($result === null) {
            return $this->jsonError($response, 404, 'STORY_NOT_FOUND', 'No story: ' . $args['uuidStory']);
        }
        $response->getBody()->write(json_encode($result));
        return $response->withHeader('Content-Type', 'application/json');
    }

    // PUT /api/admin/stories/{uuidStory}
    public function updateStory(Request $request, Response $response, array $args): Response
    {
        if (!$this->isAdmin($request)) {
            return $this->forbidden($response);
        }
        $data = $request->getParsedBody();
        if (empty($data)) {
            return $this->jsonError($response, 400, 'EMPTY_DATA', 'Request body required');
        }
        $result = $this->crudPort->updateStory($args['uuidStory'], $data);
        if ($result === null) {
            return $this->jsonError($response, 404, 'STORY_NOT_FOUND', 'No story: ' . $args['uuidStory']);
        }
        $response->getBody()->write(json_encode($result));
        return $response->withHeader('Content-Type', 'application/json');
    }

    // GET /api/admin/stories/{uuidStory}/{entityType}
    public function listEntities(Request $request, Response $response, array $args): Response
    {
        if (!$this->isAdmin($request)) {
            return $this->forbidden($response);
        }
        $result = $this->crudPort->listEntities($args['uuidStory'], $args['entityType']);
        if ($result === null) {
            return $this->jsonError($response, 404, 'STORY_NOT_FOUND', 'No story: ' . $args['uuidStory']);
        }
        $response->getBody()->write(json_encode($result));
        return $response->withHeader('Content-Type', 'application/json');
    }

    // POST /api/admin/stories/{uuidStory}/{entityType}
    public function createEntity(Request $request, Response $response, array $args): Response
    {
        if (!$this->isAdmin($request)) {
            return $this->forbidden($response);
        }
        $data = $request->getParsedBody();
        if (empty($data)) {
            return $this->jsonError($response, 400, 'EMPTY_DATA', 'Request body required');
        }
        try {
            $result = $this->crudPort->createEntity($args['uuidStory'], $args['entityType'], $data);
        } catch (\Games\Paths\Core\Domain\Story\StoryValidationException $e) {
            return $this->validationError($response, $e);
        }
        if ($result === null) {
            return $this->jsonError($response, 404, 'STORY_NOT_FOUND', 'No story: ' . $args['uuidStory']);
        }
        $response->getBody()->write(json_encode($result));
        return $response->withStatus(201)->withHeader('Content-Type', 'application/json');
    }

    // GET /api/admin/stories/{uuidStory}/{entityType}/{entityUuid}
    public function getEntity(Request $request, Response $response, array $args): Response
    {
        if (!$this->isAdmin($request)) {
            return $this->forbidden($response);
        }
        $result = $this->crudPort->getEntity($args['uuidStory'], $args['entityType'], $args['entityUuid']);
        if ($result === null) {
            return $this->jsonError($response, 404, 'ENTITY_NOT_FOUND', 'Not found: ' . $args['entityUuid']);
        }
        $response->getBody()->write(json_encode($result));
        return $response->withHeader('Content-Type', 'application/json');
    }

    // PUT /api/admin/stories/{uuidStory}/{entityType}/{entityUuid}
    public function updateEntity(Request $request, Response $response, array $args): Response
    {
        if (!$this->isAdmin($request)) {
            return $this->forbidden($response);
        }
        $data = $request->getParsedBody();
        if (empty($data)) {
            return $this->jsonError($response, 400, 'EMPTY_DATA', 'Request body required');
        }
        try {
            $result = $this->crudPort->updateEntity($args['uuidStory'], $args['entityType'], $args['entityUuid'], $data);
        } catch (\Games\Paths\Core\Domain\Story\StoryValidationException $e) {
            return $this->validationError($response, $e);
        }
        if ($result === null) {
            return $this->jsonError($response, 404, 'ENTITY_NOT_FOUND', 'Not found: ' . $args['entityUuid']);
        }
        $response->getBody()->write(json_encode($result));
        return $response->withHeader('Content-Type', 'application/json');
    }

    private function validationError(Response $response, \Games\Paths\Core\Domain\Story\StoryValidationException $e): Response
    {
        $report = $e->getReport();
        $response->getBody()->write(json_encode([
            'error' => 'INVALID_STORY',
            'message' => $report->summary(),
            'errors' => array_map(fn ($err) => $err->toArray(), $report->getErrors()),
        ]));
        return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
    }

    // DELETE /api/admin/stories/{uuidStory}/{entityType}/{entityUuid}
    public function deleteEntity(Request $request, Response $response, array $args): Response
    {
        if (!$this->isAdmin($request)) {
            return $this->forbidden($response);
        }
        $deleted = $this->crudPort->deleteEntity($args['uuidStory'], $args['entityType'], $args['entityUuid']);
        if (!$deleted) {
            return $this->jsonError($response, 404, 'ENTITY_NOT_FOUND', 'Not found: ' . $args['entityUuid']);
        }
        $response->getBody()->write(json_encode([
            'status' => 'DELETED',
            'uuid' => $args['entityUuid'],
            'entityType' => $args['entityType'],
        ]));
        return $response->withHeader('Content-Type', 'application/json');
    }

    private function isAdmin(Request $request): bool
    {
        $role = $request->getAttribute('role');
        return $role === 'ADMIN';
    }

    private function forbidden(Response $response): Response
    {
        $response->getBody()->write(json_encode([
            'error' => 'FORBIDDEN',
            'message' => 'Insufficient permissions (ADMIN role required)',
        ]));
        return $response->withStatus(403)->withHeader('Content-Type', 'application/json');
    }

    private function jsonError(Response $response, int $status, string $error, string $message): Response
    {
        $response->getBody()->write(json_encode([
            'error' => $error,
            'message' => $message,
        ]));
        return $response->withStatus($status)->withHeader('Content-Type', 'application/json');
    }
}
